src/webgpu/shader/validation/expression/call/builtin/textureGather.spec.ts

```diff
@@ -0,0 +1,333 @@
+    hasComponentArg: true,
+    coordsArgType: Type.vec3f,
+    hasArrayIndexArg: true,
+  },
+  texture_depth_2d: { coordsArgType: Type.vec2f, offsetArgType: Type.vec2i },
+  texture_depth_2d_array: {
+    coordsArgType: Type.vec2f,
+    hasArrayIndexArg: true,
+    offsetArgType: Type.vec2i,
+  },
+  texture_depth_cube: { coordsArgType: Type.vec3f },
+  texture_depth_cube_array: { coordsArgType: Type.vec3f, hasArrayIndexArg: true },
+} as const;
+
+const kTextureTypes = keysOf(kValidTextureGatherParameterTypes);
+const kValuesTypes = objectsToRecord(kAllScalarsAndVectors);
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+g.test('component_argument')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#texturegather')
+  .desc(
+    `
+Validates that only incorrect components arguments are rejected by ${builtin}
+`
+  )
+  .params(u =>
+    u
+      .combine('textureType', keysOf(kValidTextureGatherParameterTypes))
+      // filter out types with no component argument
+      .filter(t => !!kValidTextureGatherParameterTypes[t.textureType].hasComponentArg)
+      .combine('componentType', keysOf(kValuesTypes))
+      .beginSubcases()
+      .combine('value', [-1, 0, 1, 2, 3, 4] as const)
+      // filter out unsigned types with negative values
+      .filter(t => !isUnsignedType(kValuesTypes[t.componentType]) || t.value >= 0)
+      .expand('offset', t =>
+        kValidTextureGatherParameterTypes[t.textureType].offsetArgType ? [false, true] : [false]
+      )
+  )
+  .fn(t => {
+    const { textureType, componentType, offset, value } = t.params;
+    const componentArgType = kValuesTypes[componentType];
+    const { offsetArgType, coordsArgType, hasArrayIndexArg } =
+      kValidTextureGatherParameterTypes[textureType];
+
+    const componentWGSL = componentArgType.create(value).wgsl();
+    const coordWGSL = coordsArgType.create(0).wgsl();
+    const arrayWGSL = hasArrayIndexArg ? ', 0' : '';
+    const offsetWGSL = offset ? `, ${offsetArgType?.create(0).wgsl()}` : '';
+
+    const code = `
+@group(0) @binding(0) var s: sampler;
+@group(0) @binding(1) var t: ${textureType};
+@fragment fn fs() -> @location(0) vec4f {
+  let v = textureGather(${componentWGSL}, t, s, ${coordWGSL}${arrayWGSL}${offsetWGSL});
+  return vec4f(0);
+}
+`;
+    const expectSuccess =
+      (isConvertible(componentArgType, Type.i32) || isConvertible(componentArgType, Type.u32)) &&
+      value >= 0 &&
+      value <= 3;
+    t.expectCompileResult(expectSuccess, code);
+  });
+
+g.test('component_argument,non_const')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#texturegather')
+  .desc(
+    `
+Validates that only non-const components arguments are rejected by ${builtin}
+`
+  )
+  .params(u =>
+    u
+      .combine('textureType', keysOf(kValidTextureGatherParameterTypes))
+      // filter out types with no component argument
+      .filter(t => !!kValidTextureGatherParameterTypes[t.textureType].hasComponentArg)
+      .combine('varType', ['c', 'u', 'l'])
```

Comment:
I don't see a declaration for `u` so it will fail validation, but for the wrong reason.

Reply:
I'd also consider making this an expression (e.g. `c + 1`).

Reply:
`u` was missing. Added. Not sure what you mean by `c + 1`.

Reply:
Making it an expression instead of just a reference to a declaration.

Reply:
I'm not following. Why does need to be an expression? The thing being tested is does `c` (a const) work and `u` a uniform and `l` an non-constant, fail. It's testing that. What does `c + 1` change about that?

Reply:
To me, it tests that an expression is acceptable and not just a name.

Reply:
constant expressions are already tested by the other test though right?

I think I'm confused. We're testing in other tests which types are acceptable. Expressions evaluate to a type so testing expressions is redundant. Instead, in some other part of the CTS, that each expression evaluates to the correct type would be tested. Then, here, all that's tested is (1) that only the correct types are allowed (that's the previous test to this one) and (2) the extra restrictions (in this case that this parameter must be const).

Otherwise, I'm not sure why would wouldn't have to test every possible expression `c + 1`, `c - 1` (because maybe minus converts to a different type, `someFunc(c)`, `1 - c`, ` + c`, etc, etc.

Reply:
If they are covered elsewhere that's ok then.